compositor_render/src/transformations/web_renderer/transformation.rs

```diff
@@ -0,0 +1,176 @@
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    env,
+    error::Error,
+    io,
+    num::ParseIntError,
+    process::{self, Command},
+    rc::Rc,
+};
+
+use compositor_common::scene::{NodeId, Resolution, TransformationRegistryKey};
+use image::ImageError;
+use log::warn;
+
+use crate::renderer::{
+    texture::Texture,
+    transformation::{Transformation, TransformationParams},
+    WgpuCtx,
+};
+
+use super::{
+    electron_api::{ElectronApi, ElectronApiError},
+    SessionId, Url,
+};
+
+pub struct WebRenderer {
+    api: ElectronApi,
+    wgpu_ctx: Rc<WgpuCtx>,
+    renderer_process: process::Child,
+    session_ids: RefCell<HashMap<Url, SessionId>>,
+}
+
+impl Transformation for WebRenderer {
+    fn apply(
+        &self,
+        params: &TransformationParams,
+        _sources: &HashMap<NodeId, Texture>,
+        target: &Texture,
+    ) -> Result<(), Box<dyn Error>> {
+        let TransformationParams::String(url) = params else {
+            return Err(Box::new(WebRendererApplyError::InvalidParam));
+        };
+
+        let mut session_ids = self.session_ids.borrow_mut();
+        let session_id = match session_ids.get(url) {
+            Some(id) => id.to_owned(),
```

Comment:
Do you need to clone it? Wouldn't a borrowing session id be enough?

Also if you have a case where you need to clone string it might be more efficient to use Arc<str>/Rc<str> instead. I did not benchmark this, but I suspect creating ref counted pointer is cheaper than allocating memorry